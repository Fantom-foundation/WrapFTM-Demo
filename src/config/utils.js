export function envToBool(value) {
    return value === 'true' || value === '1';
}
