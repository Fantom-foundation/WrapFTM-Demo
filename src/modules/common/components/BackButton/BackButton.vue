<script setup>
import { FButton, useMethods } from 'fantom-vue3-components';
import AppIconset from '@/modules/common/components/AppIconset/AppIconset.vue';

const props = defineProps({
    viewSwitcherId: {
        type: String,
        default: '',
    },
    componentId: {
        type: String,
        default: '',
    },
});

const emit = defineEmits(['back-button-click']);

function onBackButtonClick() {
    if (props.viewSwitcherId && props.componentId) {
        const { goBack } = useMethods(props.viewSwitcherId).getMethods();

        if (goBack) {
            goBack(props.componentId);
        }
    }

    emit('back-button-click');
}
</script>

<template>
    <FButton tertiary round :title="$t('common.goBack')" @click="onBackButtonClick">
        <AppIconset icon="arrowLeft" size="16px" class="fsvgicon-rtl-mirror" />
    </FButton>
</template>
