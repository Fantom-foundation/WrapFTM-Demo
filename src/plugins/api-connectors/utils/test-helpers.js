export class MyApiConnectorMock {
    __errors__ = {};
    __results__ = {};

    async processQuery() {
        return new Promise((resolve, reject) => {
            // console.log(this.__errors__.processQuery);
            if (this.__errors__.processQuery !== undefined) {
                reject(this.__errors__.processQuery);
            } else if (this.__results__.processQuery !== undefined) {
                resolve(this.__results__.processQuery);
            } else {
                resolve('result');
            }
        });
    }

    async processMutation() {
        return new Promise((resolve, reject) => {
            // console.log(this.__errors__.processQuery);
            if (this.__errors__.processMutation !== undefined) {
                reject(this.__errors__.processMutation);
            } else if (this.__results__.processMutation !== undefined) {
                resolve(this.__results__.processMutation);
            } else {
                resolve('result');
            }
        });
    }

    processResult(result) {
        return result;
    }
}
