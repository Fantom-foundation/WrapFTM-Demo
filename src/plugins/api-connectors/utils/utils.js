import { isProxy, isReactive, isRef, toRaw, unref } from 'vue';

/**
 * @param {function[]} functions
 * @param args
 */
export function callFunctions(functions, ...args) {
    functions.forEach((fn) => {
        if (typeof fn === 'function') {
            fn(...args);
        }
    });
}

export function deepToRaw(sourceObj) {
    const objectIterator = (input) => {
        if (Array.isArray(input)) {
            return input.map((item) => objectIterator(item));
        }
        if (isRef(input)) {
            return objectIterator(unref(input));
        }
        if (isReactive(input) || isProxy(input)) {
            return objectIterator(toRaw(input));
        }
        if (input && typeof input === 'object') {
            return Object.getOwnPropertyNames(input).reduce((acc, key) => {
                acc[key] = objectIterator(input[key]);
                return acc;
            }, {});
        }
        return input;
    };

    return objectIterator(sourceObj);
}
