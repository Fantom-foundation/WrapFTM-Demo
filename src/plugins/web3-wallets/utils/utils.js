import { toInt } from '@/utils/big-number/big-number.js';
import { CHAIN_NAMES } from '@/plugins/web3-wallets/utils/chain-names.js';

export function getChainNameById(chainId) {
    const chainIdI = toInt(chainId);

    return CHAIN_NAMES[chainIdI] || '';
}
