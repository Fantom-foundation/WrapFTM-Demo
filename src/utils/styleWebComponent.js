function applyStyles(elem, styles) {
    Object.keys(styles).forEach((prop) => {
        elem.style[prop] = styles[prop];
    });
}

function findWebComponent(wcSelector) {
    const wcs = wcSelector.split('>');
    let wc = null;

    wcs.forEach((wcS) => {
        wc = (wc === null ? document : wc?.shadowRoot)?.querySelector(wcS);
    });

    return wc;
}

/**
 * @param {string} wcSelector
 * @param {Object<string, Object>} styles
 */
export function styleWebComponent(wcSelector, styles) {
    const wc = findWebComponent(wcSelector);

    if (!wc) {
        throw new Error(`Can't find web component '${wcSelector}'`);
    }

    if (!wc.shadowRoot) {
        throw new Error(`'${wcSelector}' is not a web component`);
    }

    Object.keys(styles).forEach((selector) => {
        if (selector === 'root') {
            applyStyles(wc, styles[selector]);
        } else {
            const elems = wc.shadowRoot.querySelectorAll(selector);

            if (elems.length > 0) {
                elems.forEach((elem) => {
                    applyStyles(elem, styles[selector]);
                });
            } else {
                console.warn(`Can't find element(s) '${selector}'`);
            }
        }
    });
}
