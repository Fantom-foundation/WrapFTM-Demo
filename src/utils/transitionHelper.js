/**
 * Code taken from: https://developer.chrome.com/docs/web-platform/view-transitions/
 * Current view transitions api support: https://caniuse.com/?search=view%20transitions
 *
 * @param {boolean} skipTransition
 * @param {array} classNames
 * @param {function} updateDOM
 * @return {{updateCallbackDone: Promise<unknown>, ready: Promise<never>, skipTransition: skipTransition, finished: Promise<unknown>}|*}
 */
export function transitionHelper({ skipTransition = false, classNames = [], updateDOM }) {
    if (skipTransition || !document.startViewTransition) {
        const updateCallbackDone = Promise.resolve(updateDOM()).then(() => {});
        const ready = Promise.reject(Error('View transitions unsupported'));

        // Avoid spamming the console with this error unless the promise is used.
        ready.catch(() => {});

        return {
            ready,
            updateCallbackDone,
            finished: updateCallbackDone,
            skipTransition: () => {},
        };
    }

    document.documentElement.classList.add(...classNames);

    const transition = document.startViewTransition(updateDOM);

    transition.finished.finally(() =>
        document.documentElement.classList.remove(...classNames)
    );

    return transition;
}
