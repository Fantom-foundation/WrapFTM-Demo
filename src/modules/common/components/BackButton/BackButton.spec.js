import { mount } from '@vue/test-utils';
import { destroyWrapper } from 'fantom-vue3-components/src/test/utils.js';
import BackButton from '@/modules/common/components/BackButton/BackButton.vue';
import { FViewSwitcher } from 'fantom-vue3-components';
import { markRaw, nextTick } from 'vue';

let wrapper = null;
const VIEW_SWITCHER_ID = 'switcher-id';
function APP_STRUCTURE() {
    return [
        {
            id: 'Component',
            type: 'component',
            _c: [
                {
                    id: 'ComponentWithBackButton',
                    type: 'component',
                },
            ],
        },
    ];
}

// defineComponent()
const Component = markRaw({
    name: 'Component',
    template: `<div></div>`,
});

const ComponentWithBackButton = markRaw({
    name: 'ComponentWithBackButton',
    components: { BackButton },
    template: `<div><BackButton :view-switcher-id="viewSwitcherId" component-id="ComponentWithBackButton" /></div>`,
    data() {
        return {
            viewSwitcherId: VIEW_SWITCHER_ID,
        };
    },
});

const Playground = {
    components: { FViewSwitcher, Component, ComponentWithBackButton },
    template: `
        <FViewSwitcher
            :id="viewSwitcherId"
            :app-structure="appStructure"
            :components="components"
            default-component="ComponentWithBackButton"
        />
    `,
    data() {
        return {
            viewSwitcherId: VIEW_SWITCHER_ID,
            appStructure: APP_STRUCTURE(),
            components: { Component, ComponentWithBackButton },
        };
    },
};

function createWrapper(options = {}) {
    return mount(BackButton, options);
}

function createPlayground(options = {}) {
    return mount(Playground, options);
}

afterEach(() => {
    destroyWrapper(wrapper);
});

describe('BackButton', () => {
    it('should match snapshot', () => {
        wrapper = createWrapper();

        expect(wrapper.element).toMatchSnapshot();
    });

    it('should go to parent component if "viewSwitcherId" and "componentId" props are given', async () => {
        wrapper = createPlayground();

        await wrapper.findComponent(BackButton).trigger('click');
        await nextTick();

        expect(wrapper.findComponent(Component).exists()).toBe(true);
    });

    it('should emit "back-button-click" event', async () => {
        wrapper = createWrapper();

        await wrapper.find('button').trigger('click');

        expect(wrapper.emitted('back-button-click')).toBeDefined();
    });
});
