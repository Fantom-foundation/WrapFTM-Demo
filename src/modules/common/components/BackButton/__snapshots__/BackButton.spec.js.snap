// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BackButton > should match snapshot 1`] = `
<button
  class="fbutton btn btn-tertiary btn-round btn-samesize"
  title="Go back"
  type="button"
>
  <span
    class="fbutton_label"
  >
    
    <span
      aria-hidden="true"
      class="fsvgicon fsvgicon-fill fsvgicon-rtl-mirror"
      style="width: 16px; height: 16px;"
    >
      <span
        class="fsvgicon-icon"
        style="transform: rotate();"
      >
        
        <svg
          height="15.995"
          viewBox="0 0 16 15.995"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            data-name="Component 3 â€“ 5"
            id="Component_3_5"
            transform="translate(16) rotate(90)"
          >
            <path
              d="M30.677,19.406a.985.985,0,0,0-1.395,0L23.868,24.82v-11.6a.993.993,0,1,0-1.978,0v11.5l-5.257-5.264a.993.993,0,0,0-1.4,1.4l7,7a1.009,1.009,0,0,0,1.4,0L30.677,20.8a.985.985,0,0,0,0-1.395Z"
              data-name="Path 4"
              id="Path_4"
              transform="translate(-14.971 -12.138)"
            />
          </g>
        </svg>
        
      </span>
    </span>
    
  </span>
</button>
`;
